Zend/zend_enum.c

```diff
@@ -0,0 +1,338 @@
+/*
+   +----------------------------------------------------------------------+
+   | Zend Engine                                                          |
+   +----------------------------------------------------------------------+
+   | Copyright (c) Zend Technologies Ltd. (http://www.zend.com)           |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 2.00 of the Zend license,     |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | http://www.zend.com/license/2_00.txt.                                |
+   | If you did not receive a copy of the Zend license and are unable to  |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@zend.com so we can mail you a copy immediately.              |
+   +----------------------------------------------------------------------+
+   | Authors: Ilija Tovilo <ilutov@php.net>                               |
+   +----------------------------------------------------------------------+
+*/
+
+#include "zend.h"
+#include "zend_API.h"
+#include "zend_compile.h"
+#include "zend_interfaces_arginfo.h"
+
+#define ZEND_ENUM_PROPERTY_ERROR() \
+	zend_throw_error(NULL, "Enum properties are immutable")
+
+#define ZEND_ENUM_DISALLOW_MAGIC_METHOD(propertyName, methodName) \
+	do { \
+		if (ce->propertyName) { \
+			zend_error_noreturn(E_COMPILE_ERROR, "Enum may not include %s", methodName); \
+		} \
+	} while (0);
+
+static zend_object_handlers enum_handlers;
+
+zend_object *zend_enum_new(zval *result, zend_class_entry *ce, zval *case_name_zv, zval *scalar_zv)
+{
+	// Temporarily remove the ZEND_ACC_ENUM flag to allow instantiation
+	ce->ce_flags &= ~ZEND_ACC_ENUM;
+	object_init_ex(result, ce);
+	ce->ce_flags |= ZEND_ACC_ENUM;
+
+	zend_object *zobj = Z_OBJ_P(result);
+	zobj->handlers->write_property(zobj, ZSTR_KNOWN(ZEND_STR_NAME), case_name_zv, NULL);
+	if (scalar_zv != NULL) {
+		zobj->handlers->write_property(zobj, ZSTR_KNOWN(ZEND_STR_VALUE), scalar_zv, NULL);
+	}
```

Comment:
Instead of going through write_property, it's fine to do something like

```
ZVAL_COPY(OBJ_PROP_NUM(zobj), 0), case_name_zv);
if (scalar_zv != NULL) {
    ZVAL_COPY(OBJ_PROP_NUM(zobj), 1), scalar_zv);
}
```

here, as you control the memory layout.